Т3-14 Bookings

RomanAntipenko opened the pull request:
feat: Реализуй хранения всей инфы в БД
feat: Добавь новую сущность Booking и все Дто для нее итд
feat: Реализуй новую сушность комменты
feat: методы на getItems должны теперь возвращать предметы с комментами и для владельца с lastBooking и nextBooking
feat: Настрой создание БД в schema.sql
fix: Переделай некоторые модели и ДТО для корректной работы с БД

RomanAntipenko committed:
feat: Реализуй хранения всей инфы в БД
feat: Добавь новую сущность Booking и все Дто для нее итд
feat: Реализуй новую сушность комменты
feat: методы на getItems должны теперь возвращать предметы с комментами и для владельца с lastBooking и nextBooking
feat: Настрой создание БД в schema.sql
fix: Переделай некоторые модели и ДТО для корректной работы с БД

RomanAntipenko committed:
Подправь Пом

RomanAntipenko committed:
Подправь imports

johnskyi commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -1,12 +1,59 @@
 @RequestMapping(path = "/bookings")
 public class BookingController {
+    private final BookingService bookingService;
+
```
убери пожалуйста эти пустые строки между филдами, это нарушает код стайл

johnskyi commented on src/main/java/ru/practicum/shareit/booking/dto/BookingMapper.java:
```diff
@@ -0,0 +1,39 @@
+import ru.practicum.shareit.item.model.Item;
+import ru.practicum.shareit.user.model.User;
+
+public class BookingMapper {
```
В этом классе только статические методы, значит экземпляры класса нам не нужны. Предлагаю защитить его от этого путем создания приватного контроллера без параметров. Это ожно реализовать также через ломбок или поставить аннотацию @UtilityClass

johnskyi commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -1,7 +1,35 @@
+
+@Getter
+@Setter
+@ToString
```
смотри использовать эту аннотацию нужно очень острожно, она в том числе и генерирует код для полей, которые подтянуты по связям и ленивым связям. При попытке обращении к ним может возникнуть исключение ленивой инициализации. Советую почитать об этом в статьях в гугле и исключить специальной аннотацией такие поля из toString

johnskyi commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -0,0 +1,86 @@
+import java.time.LocalDateTime;
+import java.util.Collection;
+
+public interface BookingRepository extends JpaRepository<Booking, Long> {
```
репозиторий должен быть помечен аннотацией @Repository

johnskyi commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,165 @@
+public class BookingServiceImpl implements BookingService {
+
+    private final BookingRepository bookingRepository;
+
```
пустые пол между филдами нужно убрать

johnskyi commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -1,12 +1,59 @@
 @RestController
+@RequiredArgsConstructor
 @RequestMapping(path = "/bookings")
 public class BookingController {
```
добавь пожалуйста в контроллере логирование для запросов

RomanAntipenko replied:
Мне логировать каждый запрос (тогда немного не понимаю что написать в них) или на будущее просто нужно аннотацию @sl4fj просто добавить ?

RomanAntipenko replied:
Или типо в каждом методе log.info(“вызван такой то метод контроллера”)?

johnskyi commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,165 @@
+    public Booking createBooking(long userId, BookingDto bookingDto) {
+        Optional<User> user = userRepository.findById(userId);
+        Optional<Item> item = itemRepository.findById(bookingDto.getItemId());
+        if (user.isEmpty()) {
```
чтобы не использовать такую конструкцию сложную для проверки содержимого Optional, можно воспользовать методами Optional например такими как orEslseTrow  и бросать в нем исключение. Это более современный подход. Попробуй это исправить везде

johnskyi commented on src/main/java/ru/practicum/shareit/errorhandling/ErrorHandler.java:
```diff
@@ -45,4 +50,43 @@ public ErrorResponse ownerIdMismatchHandler(final OwnerIdMismatchException e) {
         );
     }
 
+    @ExceptionHandler
```
здесь в обработчиках также необходимо добавить логирование и логировать ошибки на уровни log.debug или  log.error. Это очень облегчает траблшутинг

johnskyi commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -23,29 +24,25 @@ public class ItemController {
     public ItemDto createItem(@RequestHeader("X-Sharer-User-Id") long userId,
```
добавь пожалуйста логирование запросов

johnskyi commented on src/main/java/ru/practicum/shareit/item/dto/CommentMapper.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.time.LocalDateTime;
+
+public class CommentMapper {
```
про приватный конструктор в таком классе писал выше. Здесь та же история)

johnskyi commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -1,6 +1,7 @@
 import ru.practicum.shareit.item.model.Item;
+import ru.practicum.shareit.user.model.User;
 
 public class ItemMapper {
```
и тут аналогично верхнему комментарию

johnskyi commented on src/main/java/ru/practicum/shareit/item/model/Comment.java:
```diff
@@ -0,0 +1,30 @@
+
+@Getter
+@Setter
+@ToString
```
здесь необзодимо исключить все поля по связям из toString, подробно про это расписал выше

johnskyi commented on src/main/java/ru/practicum/shareit/item/repository/CommentRepository.java:
```diff
@@ -0,0 +1,10 @@
+import ru.practicum.shareit.item.model.Comment;
+
+import java.util.List;
+
```
@Repository

johnskyi commented on src/main/java/ru/practicum/shareit/item/repository/ItemRepository.java:
```diff
@@ -1,17 +1,16 @@
 
-public interface ItemRepository {
-    Item createItem(Item item);
+public interface ItemRepository extends JpaRepository<Item, Long> {
```
@Repository

RomanAntipenko committed:
style: Убери строки между полями
feat: Добавь аннотацию @UtilityClass в мапперы
fix: Убери toString у полей которые связывают таблицы
feat: Добавь @Repository во все репозитории
feat: Добавил логирование в Controller'ах
feat: Добавил логирование в ExceptionHandler
fix: там где Optional в получении предметов из бд поставь orElseThrow с исключением
